File: Baekjoon/6th EI/1991.java
import java.io.*;
import java.util.*;

class Node {
    char data;
    Node left;
    Node right;

    Node(char data) {
        this.data = data;
    }
}

class Tree {
    Node root;

    public void createNode(char data, char leftData, char rightData) {
        if (root == null) {
            root = new Node(data);
            if (leftData != '.') {
                root.left = new Node(leftData);
            }
            if (rightData != '.') {
                root.right = new Node(rightData);
            }
        } else {
            searchNode(root, data, leftData, rightData);
        }
    }

    public void searchNode(Node root, char data, char leftData, char rightData) {
        if (root == null) {
            return;
        } else if (root.data == data) {
            if (leftData != '.') {
                root.left = new Node(leftData);
            }
            if (rightData != '.') {
                root.right = new Node(rightData);
            }
        } else {
            searchNode(root.left, data, leftData, rightData);
            searchNode(root.right, data, leftData, rightData);
        }
    }

    // 전위순회 : 루트 -> 좌 -> 우
    public void preorder(Node root) {
        System.out.print(root.data);
        if (root.left != null)
            preorder(root.left);
        if (root.right != null)
            preorder(root.right);
    }

    // 중위순회 : 좌 -> 루트 -> 우
    public void inorder(Node root) {
        if (root.left != null)
            inorder(root.left);
        System.out.print(root.data);
        if (root.right != null)
            inorder(root.right);
    }

    // 후위순회 : 좌 -> 우 -> 루트
    public void postorder(Node root) {
        if (root.left != null)
            postorder(root.left);
        if (root.right != null)
            postorder(root.right);
        System.out.print(root.data);
    }

}

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());

        Tree tree = new Tree();

        for (int i = 0; i < N; i++) {
            char[] data;
            data = br.readLine().replaceAll(" ", "").toCharArray();
            tree.createNode(data[0], data[1], data[2]);
        }

        // 전위 순회
        tree.preorder(tree.root);
        System.out.println();

        // 중위 순회
        tree.inorder(tree.root);
        System.out.println();

        // 후위 순회
        tree.postorder(tree.root);

        br.close();
    }

}

File: 1991.kt
import java.io.*
import java.util.*

var N = 0
var a = Array<Pair<Int,Int>>(26){Pair(0,0)}

fun preorder(n:Int){
    if (n==-1) {
        return
    }
    print('A'+n)
    preorder(a[n].first)
    preorder(a[n].second)
}

fun inorder(n:Int){
    if (n==-1) {
        return
    }
    else {
        inorder(a[n].first)
        print('A'+n)
        inorder(a[n].second)
    }
}

fun postorder(n:Int){
    if (n==-1) {
        return
    }
    else {
        postorder(a[n].first)
        postorder(a[n].second)
        print('A'+n)
    }
}
fun main() {
    val br = System.`in`.bufferedReader()
    
    N = br.readLine().toInt()
    
    repeat(N){
        val carr = br.readLine().split(' ').map{it[0]}
        
        var left = 
        if (carr[1]=='.') {
            -1
        } else {
            carr[1]-'A'
        }
        
        var right = 
        if (carr[2]=='.') {
            -1
        } else {
            carr[2]-'A'
        }
        
        a[carr[0]-'A'] = Pair(left,right)
    }
    
    preorder(0)
    println()
    inorder(0)
    println()
    postorder(0)
}
